gms/gossiper.cc

```diff
@@ -1708,13 +1712,58 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
     utils::chunked_vector<application_state> changed;
     auto&& remote_map = remote_state.get_application_state_map();
 
+    auto revert_to_old_app_state = [&] {
+        try {
+            logger.warn("Started to roll back to old app state for node={}, old_state={}, status={}, replicate_failed={}, listener_failed={}",
+                    addr, old_state, get_gossip_status(addr), replicate_failed, listener_failed);
+            local_state = old_state;
+            replicate(addr, old_state).get();
```

Comment:
This can roll back a state change for which we already ran a change listener. In other words, gossip states are no longer monotonic within the process's lifetime. That can break some expectations, and is harder to reason about than monotonic changes. I would rather have us solve this problem in a different way.

Reply:
First, the failure of the handler should be a very rare event. This patch is about how to handle the worse case scenario. Yes, when we roll back, the listener may have run.  But the inconsistency is supposed to be temporary when the listener is called again. Note, currently, if the failure happens, we abort the whole process while we can probably recover, which is horrible. 

I would prefer a temporary inconsistency over a whole node and even whole cluster abort which is a much much worse. Also, a node can stay inconsistent if it does not get the gossip update temporarily too. 

You are welcome to suggest solution to handle the failure while does not abort the scylla process and not leave the node in an inconsistency state forever.

Reply:
> First, the failure of the handler should be a very rare event. This patch is about how to handle the worse case scenario. Yes, when we roll back, the listener may have run. But the inconsistency is supposed to be temporary when the listener is called again. Note, currently, if the failure happens, we abort the whole process while we can probably recover, which is horrible.
> 
> I would prefer a temporary inconsistency over a whole node and even whole cluster abort which is a much much worse. Also, a node can stay inconsistent if it does not get the gossip update temporarily too.

Inconsistency can lead to logic failure in layers above the gossiper, which can lead to worse things than loss of availability. It can lead to loss of data or incorrect query results.

> 
> You are welcome to suggest solution to handle the failure while does not abort the scylla process and not leave the node in an inconsistency state forever.

I don't think there is a general solution. It depends on the case.

In the scenario you mentioned, why does this patch solve the problem? Suppose that some nodes removed the node from gossip, and some not, at the point where it jumps to normal. Those who did, will fail in the listener and, after this patch, revert to the old state. Those who did not, will put the node into the ring. The cluster will be in an inconsistent state.

I also don't understand why we fail on missing host_id, if we don't have any states for the bootstrapping node, why aren't all states, including host id, pulled in?

Reply:
For example:

n1 is the existing node
n2 is the new bootstrapping node
n2 finishes streaming
n2 advertises NORMAL 
n1 does not receive the NORMAL updates yet for network issues
n1 removes n2 from gossip because n2 does not update gossip, fat_client  removal
n1 receives NORMAL update but all other state of n2 is removed
n1 throws in the handler because it can not find the host_id

Reply:
This patch helps the problem because after n1 rolls back, n1 has a chance to talk with n2 and get all the states from n2 again and run the normal state handler successfully.

Reply:
Again, the inconsistency is temporary. The general statement inconsistency can cause data inconsistency is true. But this can happen if a node is partitioned and does not know the new node is in NORMAL status too. This patch is not making it worse.

Reply:
Gossip verbs are one way, so 

> I'm still interested in [#9658 (comment)](https://github.com/scylladb/scylla/pull/9658#discussion_r765690600), which may be the culprit of the issue which motivated this. But this PR can be considered separately.

Yes, this patch handles more generic problems: what happens when a gossip handler fails. The discussion there is just one of the reasons how the handler can fail.

Reply:
Gossip messages are async and one way message. Removal can not be mutually exclusive with what you called syns. Because we do not know when the diff would come and can not wait for it forever which will block gossip runs. 
```
1) 
take the lock
find the gossip state diff and ask the peer to send diff
the diff never arrives

2) 
can not take the lock to remove a node
```

Reply:
Is it possible to fix with the current protocol?

Reply:
This will change the verbs which will introduce incompatibility which is a much bigger issue.

Reply:
I opened https://github.com/scylladb/scylla/issues/9803 for this